File: llm_anthology/adapters/claude.py
"""Claude native account-export (claude.ai Data Export) -> IR.

Schema (probed from the real exports on disk):
  conversation: uuid, name, summary, created_at, updated_at, account, chat_messages[]
  message:      uuid, parent_message_uuid, sender(human|assistant), created_at,
                content[], text (flattened), attachments[], files[]
  content item: type in {text, thinking, tool_use, tool_result, token_budget, ...}
    text        -> text, citations[]
    thinking    -> thinking, summaries[], (thinking_)hidden
    tool_use    -> name, input, display_content, integration_name, icon_name, id
    tool_result -> name, content, display_content, is_error, integration_name, icon_name, tool_use_id
    token_budget-> hidden (dropped)
  attachment:   file_name, file_type, file_size, extracted_content (embedded doc text)
  file:         file_name, file_uuid (NO bytes in export -> rendered as a name chip)

Branches: messages form a tree via parent_message_uuid (91 real branch points on
disk). We render the ACTIVE (latest) path — at each node choose the child with the
max created_at — and annotate a turn with {index,total} when its parent had siblings.
"""
from llm_anthology import ir

_HIDDEN_CONTENT_TYPES = {"token_budget"}


def _s(x):
    """Coerce a display field to str. Claude's display_content is sometimes a
    structured list, not a string; the real content still lives in data{}."""
    return x if isinstance(x, str) else ""


def parse_export(data):
    """A whole export file (array of conversations, or a single dict) -> [Conversation]."""
    convs = data if isinstance(data, list) else [data]
    return [parse_conversation(c) for c in convs]


def is_design_chat(data):
    """A Claude 'design chat' (design_chats/*.json) rather than a conversations.json
    entry: messages[] with `role` + a content DICT, instead of chat_messages[] with
    `sender` + a content LIST."""
    return (isinstance(data, dict) and isinstance(data.get("messages"), list)
            and "chat_messages" not in data)


def parse_design_chat(data):
    """design_chats/*.json -> IR. A distinct shape, so parse_conversation cannot read
    it: feeding one through the normal path yields a silently EMPTY conversation."""
    turns = []
    for m in (data.get("messages") or []):
        if not isinstance(m, dict):
            continue
        role = "human" if _s(m.get("role")).lower() in ("user", "human") else "assistant"
        blocks = _design_blocks(m.get("content"))
        if blocks:
            turns.append(ir.Turn(role=role, blocks=blocks, uuid=_s(m.get("uuid")),
                                 timestamp=_s(m.get("created_at"))))
    return ir.Conversation(
        id=_s(data.get("uuid")),
        title=_s(data.get("title")) or "(untitled design chat)",
        provider="claude",
        turns=turns,
        created_at=_s(data.get("created_at")),
        updated_at=_s(data.get("updated_at")),
        meta={"kind": "design_chat"},
    )


def _design_blocks(content):
    """Text lives in content.contentBlocks[] when present, else the flat
    content.content string (which duplicates the blocks)."""
    if isinstance(content, str):
        return [ir.Block("text", text=content)] if content.strip() else []
    if not isinstance(content, dict):
        return []
    blocks = []
    for b in (content.get("contentBlocks") or []):
        if not isinstance(b, dict):
            continue
        btype = _s(b.get("type"))
        if btype == "text" and _s(b.get("text")).strip():
            blocks.append(ir.Block("text", text=_s(b.get("text"))))
        elif btype == "tool_call" and isinstance(b.get("toolCall"), dict):
            blocks.extend(_tool_call_blocks(b["toolCall"]))
        else:                                   # never silently drop an unknown block
            blocks.append(ir.Block("unknown", data={"orig_type": btype, "x_raw": b}))
    if blocks:
        return blocks
    flat = _s(content.get("content"))
    return [ir.Block("text", text=flat)] if flat.strip() else []


def _tool_call_blocks(tc):
    """A design-chat tool_call {id,name,input,output} -> tool_use (+ tool_result when it
    carries an output), so it renders like the regular chat_messages tool path rather
    than as a raw payload. 140 such blocks sit in the live corpus (2 design chats)."""
    out = [ir.Block("tool_use", text="",
                    data={"name": _s(tc.get("name")), "input": tc.get("input"),
                          "id": _s(tc.get("id"))})]
    output = tc.get("output")
    if _s(output).strip():
        out.append(ir.Block("tool_result", text="",
                            data={"name": _s(tc.get("name")), "content": output,
                                  "tool_use_id": _s(tc.get("id"))}))
    return out


def parse_conversation(conv):
    messages = conv.get("chat_messages") or []
    turns = []
    for msg, branch in _active_path(messages):
        role = "human" if msg.get("sender") == "human" else "assistant"
        turns.append(ir.Turn(
            role=role,
            blocks=_blocks_from_message(msg),
            uuid=msg.get("uuid") or "",
            timestamp=msg.get("created_at") or "",
            branch=branch,
        ))
    return ir.Conversation(
        id=conv.get("uuid") or "",
        title=conv.get("name") or "(untitled)",
        provider="claude",
        turns=turns,
        created_at=conv.get("created_at") or "",
        updated_at=conv.get("updated_at") or "",
        account=_account_str(conv.get("account")),
        meta={"summary": conv.get("summary") or ""},
    )


def _account_str(acc):
    if isinstance(acc, dict):
        return acc.get("uuid") or acc.get("email_address") or acc.get("email") or ""
    return acc or ""


def _ts(m):
    return m.get("created_at") or ""


def _subtree_max_ts(messages, children):
    """Newest created_at anywhere inside each message's subtree.

    Iterative (a long chain would blow Python's recursion limit) and cycle-safe.
    Needed because the LOCALLY newest child is often an abandoned regeneration,
    while the live thread continues under an older sibling.
    """
    memo = {}
    for start in messages:
        if start.get("uuid") in memo:
            continue
        stack, onpath = [(start, False)], set()
        while stack:
            m, expanded = stack.pop()
            u = m.get("uuid")
            if expanded:
                best = _ts(m)
                for k in children.get(u, []):
                    v = memo.get(k.get("uuid"))
                    if v is not None and v > best:
                        best = v
                memo[u] = best
                onpath.discard(u)
                continue
            if u in memo or u in onpath:
                continue
            onpath.add(u)
            stack.append((m, True))
            for k in children.get(u, []):
                if k.get("uuid") not in memo:
                    stack.append((k, False))
    return memo


def _active_path(messages):
    """Walk root -> live leaf. Returns [(message, branch_or_None), ...]."""
    by_id = {m.get("uuid"): m for m in messages}
    children = {}
    for m in messages:
        children.setdefault(m.get("parent_message_uuid"), []).append(m)

    roots = [m for m in messages
             if m.get("parent_message_uuid") is None or m.get("parent_message_uuid") not in by_id]
    submax = _subtree_max_ts(messages, children)
    path, seen = [], set()
    # Walk EVERY root, chronologically. An orphaned parent link starts a second
    # root whose subtree is still real conversation content — dropping it loses
    # messages outright (measured: 42 messages / 18 conversations on real data).
    for root in sorted(roots, key=_ts):
        cur = root
        while cur is not None and cur.get("uuid") not in seen:
            seen.add(cur.get("uuid"))
            sibs = children.get(cur.get("parent_message_uuid"), [])
            branch = None
            if len(sibs) > 1:
                ordered = sorted(sibs, key=_ts)
                branch = {"index": ordered.index(cur) + 1, "total": len(sibs)}
            path.append((cur, branch))
            kids = children.get(cur.get("uuid"), [])
            # descend toward the subtree holding the globally newest message,
            # NOT merely the newest immediate child (that abandons live threads)
            cur = max(kids, key=lambda k: (submax.get(k.get("uuid"), _ts(k)), _ts(k))) if kids else None

    # Sweep ONLY nodes reachable from no root at all (e.g. an orphaned cycle).
    # Branch siblings are also "unvisited" but are intentionally off the active
    # path — sweeping those in would splice abandoned regenerations into the text.
    reachable, stack = set(), list(roots)
    while stack:
        m = stack.pop()
        u = m.get("uuid")
        if u in reachable:            # pragma: no cover - defensive: each message has a
            continue                  # single parent, so BFS cannot revisit a node
        reachable.add(u)
        stack.extend(children.get(u, []))
    orphans = [m for m in messages
               if m.get("uuid") not in reachable and m.get("uuid") not in seen]
    for m in sorted(orphans, key=_ts):
        if m.get("uuid") in seen:     # pragma: no cover - orphans are pre-filtered `not in seen`
            continue
        seen.add(m.get("uuid"))
        path.append((m, None))
    return path


def _blocks_from_message(m):
    blocks = []
    # uploaded files/attachments render above the message text in claude.ai
    for a in (m.get("attachments") or []):
        blocks.append(ir.Block(
            "attachment", text=a.get("file_name") or "attachment",
            data={"file_name": a.get("file_name"), "file_type": a.get("file_type"),
                  "file_size": a.get("file_size"), "extracted_content": a.get("extracted_content")}))
    for f in (m.get("files") or []):
        blocks.append(ir.Block(
            "file", text=f.get("file_name") or "file",
            data={"file_name": f.get("file_name"), "file_uuid": f.get("file_uuid")}))

    for item in (m.get("content") or []):
        t = item.get("type")
        if t in _HIDDEN_CONTENT_TYPES:
            continue
        if t == "text":
            blocks.append(ir.Block("text", text=_s(item.get("text")),
                                   citations=item.get("citations") or []))
        elif t == "thinking":
            blocks.append(ir.Block("thinking", text=_s(item.get("thinking")),
                                   data={"hidden": bool(item.get("thinking_hidden") or item.get("hidden")),
                                         "summaries": item.get("summaries") or []}))
        elif t == "tool_use":
            blocks.append(ir.Block("tool_use", text=_s(item.get("display_content")),
                                   data={"name": item.get("name") or "", "input": item.get("input"),
                                         "integration_name": item.get("integration_name"),
                                         "icon_name": item.get("icon_name"), "id": item.get("id")}))
        elif t == "tool_result":
            blocks.append(ir.Block("tool_result", text=_s(item.get("display_content")),
                                   data={"name": item.get("name") or "", "content": item.get("content"),
                                         "is_error": bool(item.get("is_error")),
                                         "integration_name": item.get("integration_name"),
                                         "icon_name": item.get("icon_name"),
                                         "tool_use_id": item.get("tool_use_id")}))
        else:
            # never silently drop an unknown block — pass it through for later handling
            blocks.append(ir.Block("unknown", text="", data={"orig_type": t, "x_raw": item}))
    return blocks

File: js/src/adapters/claude.ts
/**
 * Claude native-export -> IR. 1:1 with llm_anthology/adapters/claude.py.
 *
 * Claude stores a message TREE (parent_message_uuid), and regenerations create
 * sibling branches. The active-path walk descends toward the subtree holding the
 * GLOBALLY newest message, not merely the newest immediate child — the latter
 * abandons live threads. It walks EVERY root (an orphaned parent link starts a
 * second real root) and sweeps in nodes reachable from no root. On real data the
 * naive walk dropped 42 messages across 18 conversations.
 *
 * design_chats/*.json are a DIFFERENT shape (messages[] with role + a content
 * dict, not chat_messages[] with sender + a content list); parseConversation would
 * yield a silently EMPTY conversation, so isDesignChat/parseDesignChat handle them.
 */
import type { Block, Branch, Citation, Conversation } from "../ir.js";
import { block, conversation, turn } from "../ir.js";

const HIDDEN_CONTENT_TYPES = new Set(["token_budget"]);

type Rec = Record<string, unknown>;

/** Coerce a display field to str. Claude's display_content is sometimes a
 *  structured list, not a string; the real content still lives in data{}. */
function s(x: unknown): string {
  return typeof x === "string" ? x : "";
}

function get(o: unknown, k: string): unknown {
  return o !== null && typeof o === "object" ? (o as Rec)[k] : undefined;
}

/** Python truthiness for the `a or b or ""` idiom over export fields. */
function or(...vals: unknown[]): string {
  for (const v of vals) if (v) return typeof v === "string" ? v : String(v);
  return "";
}

export function parseExport(data: unknown): Conversation[] {
  const convs = Array.isArray(data) ? data : [data];
  return convs.map(parseConversation);
}

export function isDesignChat(data: unknown): boolean {
  return (
    data !== null &&
    typeof data === "object" &&
    Array.isArray((data as Rec)["messages"]) &&
    !("chat_messages" in (data as Rec))
  );
}

export function parseConversation(conv: unknown): Conversation {
  const messages = (get(conv, "chat_messages") as Rec[]) || [];
  const turns = [];
  for (const [msg, branch] of activePath(messages)) {
    const role = get(msg, "sender") === "human" ? "human" : "assistant";
    turns.push(
      turn(role, blocksFromMessage(msg), {
        uuid: or(get(msg, "uuid")),
        timestamp: or(get(msg, "created_at")),
        branch,
      }),
    );
  }
  return conversation(or(get(conv, "uuid")), or(get(conv, "name")) || "(untitled)", "claude", {
    turns,
    created_at: or(get(conv, "created_at")),
    updated_at: or(get(conv, "updated_at")),
    account: accountStr(get(conv, "account")),
    meta: { summary: or(get(conv, "summary")) },
  });
}

function accountStr(acc: unknown): string {
  if (acc !== null && typeof acc === "object") {
    return or(get(acc, "uuid"), get(acc, "email_address"), get(acc, "email"));
  }
  return typeof acc === "string" ? acc : "";
}

function ts(m: unknown): string {
  return or(get(m, "created_at"));
}

/**
 * Newest created_at anywhere inside each message's subtree. Iterative (a long
 * chain would blow the recursion limit) and cycle-safe. Needed because the LOCALLY
 * newest child is often an abandoned regeneration while the live thread continues
 * under an older sibling.
 */
function subtreeMaxTs(messages: Rec[], children: Map<unknown, Rec[]>): Map<unknown, string> {
  const memo = new Map<unknown, string>();
  for (const start of messages) {
    if (memo.has(get(start, "uuid"))) continue;
    const stack: Array<[Rec, boolean]> = [[start, false]];
    const onpath = new Set<unknown>();
    while (stack.length) {
      const [m, expanded] = stack.pop()!;
      const u = get(m, "uuid");
      if (expanded) {
        let best = ts(m);
        for (const k of children.get(u) ?? []) {
          const v = memo.get(get(k, "uuid"));
          if (v !== undefined && v > best) best = v;
        }
        memo.set(u, best);
        onpath.delete(u);
        continue;
      }
      if (memo.has(u) || onpath.has(u)) continue;
      onpath.add(u);
      stack.push([m, true]);
      for (const k of children.get(u) ?? []) {
        if (!memo.has(get(k, "uuid"))) stack.push([k, false]);
      }
    }
  }
  return memo;
}

/** A STABLE sort by ts() — Python's sorted() is stable and the active-path branch
 *  indices depend on siblings keeping their input order among equal timestamps. */
function sortedByTs<T>(arr: T[], key: (x: T) => string): T[] {
  return arr
    .map((v, i) => [v, i] as const)
    .sort((a, b) => {
      const ka = key(a[0]);
      const kb = key(b[0]);
      return ka < kb ? -1 : ka > kb ? 1 : a[1] - b[1];
    })
    .map(([v]) => v);
}

function activePath(messages: Rec[]): Array<[Rec, Branch | null]> {
  const byId = new Map<unknown, Rec>();
  for (const m of messages) byId.set(get(m, "uuid"), m);
  const children = new Map<unknown, Rec[]>();
  for (const m of messages) {
    const p = get(m, "parent_message_uuid");
    if (!children.has(p)) children.set(p, []);
    children.get(p)!.push(m);
  }

  const roots = messages.filter((m) => {
    const p = get(m, "parent_message_uuid");
    return p === null || p === undefined || !byId.has(p);
  });
  const submax = subtreeMaxTs(messages, children);
  const path: Array<[Rec, Branch | null]> = [];
  const seen = new Set<unknown>();

  // Walk EVERY root, chronologically. An orphaned parent link starts a second root
  // whose subtree is still real conversation content.
  for (const root of sortedByTs(roots, ts)) {
    let cur: Rec | null = root;
    while (cur !== null && !seen.has(get(cur, "uuid"))) {
      seen.add(get(cur, "uuid"));
      const sibs = children.get(get(cur, "parent_message_uuid")) ?? [];
      let branch: Branch | null = null;
      if (sibs.length > 1) {
        const ordered = sortedByTs(sibs, ts);
        branch = { index: ordered.indexOf(cur) + 1, total: sibs.length };
      }
      path.push([cur, branch]);
      const kids: Rec[] = children.get(get(cur, "uuid")) ?? [];
      // descend toward the subtree holding the globally newest message, NOT merely
      // the newest immediate child (that abandons live threads)
      cur = kids.length ? argmaxKid(kids, submax) : null;
    }
  }

  // Sweep ONLY nodes reachable from no root at all (e.g. an orphaned cycle). Branch
  // siblings are also "unvisited" but intentionally off the active path.
  const reachable = new Set<unknown>();
  const stack = [...roots];
  while (stack.length) {
    const m = stack.pop()!;
    const u = get(m, "uuid");
    if (reachable.has(u)) continue;
    reachable.add(u);
    stack.push(...(children.get(u) ?? []));
  }
  const orphans = messages.filter(
    (m) => !reachable.has(get(m, "uuid")) && !seen.has(get(m, "uuid")),
  );
  for (const m of sortedByTs(orphans, ts)) {
    // Duplicate UUIDs can both pass the pre-filter before the first is added to seen.
    if (seen.has(get(m, "uuid"))) continue;
    seen.add(get(m, "uuid"));
    path.push([m, null]);
  }
  return path;
}

/** max(kids, key=lambda k: (submax[k] or ts(k), ts(k))) with Python tie-breaking:
 *  on equal keys `max` keeps the FIRST, so iterate forward and use strict >. */
function argmaxKid(kids: Rec[], submax: Map<unknown, string>): Rec {
  let best = kids[0]!;
  let bestKey: [string, string] = [submax.get(get(best, "uuid")) ?? ts(best), ts(best)];
  for (let i = 1; i < kids.length; i++) {
    const k = kids[i]!;
    const key: [string, string] = [submax.get(get(k, "uuid")) ?? ts(k), ts(k)];
    if (key[0] > bestKey[0] || (key[0] === bestKey[0] && key[1] > bestKey[1])) {
      best = k;
      bestKey = key;
    }
  }
  return best;
}

function blocksFromMessage(m: Rec): Block[] {
  const blocks: Block[] = [];
  // uploaded files/attachments render above the message text in claude.ai
  for (const a of (get(m, "attachments") as Rec[]) || []) {
    blocks.push(
      block("attachment", {
        text: or(get(a, "file_name")) || "attachment",
        data: {
          file_name: get(a, "file_name"),
          file_type: get(a, "file_type"),
          file_size: get(a, "file_size"),
          extracted_content: get(a, "extracted_content"),
        },
      }),
    );
  }
  for (const f of (get(m, "files") as Rec[]) || []) {
    blocks.push(
      block("file", {
        text: or(get(f, "file_name")) || "file",
        data: { file_name: get(f, "file_name"), file_uuid: get(f, "file_uuid") },
      }),
    );
  }

  for (const item of (get(m, "content") as Rec[]) || []) {
    const t = get(item, "type");
    if (typeof t === "string" && HIDDEN_CONTENT_TYPES.has(t)) continue;
    if (t === "text") {
      blocks.push(
        block("text", {
          text: s(get(item, "text")),
          citations: (get(item, "citations") as Citation[]) || [],
        }),
      );
    } else if (t === "thinking") {
      blocks.push(
        block("thinking", {
          text: s(get(item, "thinking")),
          data: {
            hidden: Boolean(get(item, "thinking_hidden") || get(item, "hidden")),
            summaries: get(item, "summaries") || [],
          },
        }),
      );
    } else if (t === "tool_use") {
      blocks.push(
        block("tool_use", {
          text: s(get(item, "display_content")),
          data: {
            name: or(get(item, "name")),
            input: get(item, "input"),
            integration_name: get(item, "integration_name"),
            icon_name: get(item, "icon_name"),
            id: get(item, "id"),
          },
        }),
      );
    } else if (t === "tool_result") {
      blocks.push(
        block("tool_result", {
          text: s(get(item, "display_content")),
          data: {
            name: or(get(item, "name")),
            content: get(item, "content"),
            is_error: Boolean(get(item, "is_error")),
            integration_name: get(item, "integration_name"),
            icon_name: get(item, "icon_name"),
            tool_use_id: get(item, "tool_use_id"),
          },
        }),
      );
    } else {
      // never silently drop an unknown block — pass it through for later handling
      blocks.push(block("unknown", { data: { orig_type: t ?? null, x_raw: item } }));
    }
  }
  return blocks;
}

export function parseDesignChat(data: Rec): Conversation {
  const turns = [];
  for (const m of (get(data, "messages") as Rec[]) || []) {
    if (m === null || typeof m !== "object") continue;
    const role = ["user", "human"].includes(s(get(m, "role")).toLowerCase()) ? "human" : "assistant";
    const blocks = designBlocks(get(m, "content"));
    if (blocks.length) {
      turns.push(turn(role, blocks, { uuid: s(get(m, "uuid")), timestamp: s(get(m, "created_at")) }));
    }
  }
  return conversation(s(get(data, "uuid")), s(get(data, "title")) || "(untitled design chat)", "claude", {
    turns,
    created_at: s(get(data, "created_at")),
    updated_at: s(get(data, "updated_at")),
    meta: { kind: "design_chat" },
  });
}

function designBlocks(content: unknown): Block[] {
  if (typeof content === "string") {
    return content.trim() ? [block("text", { text: content })] : [];
  }
  if (content === null || typeof content !== "object") return [];
  const blocks: Block[] = [];
  for (const b of (get(content, "contentBlocks") as Rec[]) || []) {
    if (b === null || typeof b !== "object") continue;
    if (s(get(b, "type")) === "text" && s(get(b, "text")).trim()) {
      blocks.push(block("text", { text: s(get(b, "text")) }));
    } else {
      // never silently drop an unknown block
      blocks.push(block("unknown", { data: { orig_type: s(get(b, "type")), x_raw: b } }));
    }
  }
  if (blocks.length) return blocks;
  const flat = s(get(content, "content"));
  return flat.trim() ? [block("text", { text: flat })] : [];
}
